R/epidist_db.R

```diff
@@ -120,96 +127,436 @@ epidist_db <- function(disease,
+  multi_epidist <- suppressMessages(lapply(params_json, .format_epidist))
+
+  # create and return <multi_epidist> class
+  structure(
+    multi_epidist,
+    class = "multi_epidist"
+  )
+}
+
+#' Filter a list of `<epidist>` objects by disease and epi distribution
+#'
+#' @param multi_epidist A list of `<epidist>` objects.
+#' @inheritParams epidist_db
+#'
+#' @return A list of `<epidist>` objects.
+#' @keywords internal
+#' @noRd
+.filter_epidist_db <- function(multi_epidist, disease, epi_dist) {
+  # copy of user input
+  disease_ <- disease
+  epi_dist_ <- epi_dist
+
+  # clean input strings
+  disease <- clean_disease(disease)
+  epi_dist <- clean_epi_dist(epi_dist)
+
+  # get valid options from db
+  disease_db <- vapply(
+    multi_epidist, function(x) x$disease$disease,
+    FUN.VALUE = character(1)
+  )
+  epi_dist_db <- vapply(
+    multi_epidist, function(x) x$epi_dist,
+    FUN.VALUE = character(1)
+  )
+  disease_db <- c("all", clean_disease(unique(disease_db)))
+  epi_dist_db <- c("all", clean_epi_dist(unique(epi_dist_db)))
+
+  # partial matching and custom error msg
+  tryCatch(
+    error = function(cnd) {
+      disease_str <- ifelse(
+        test = disease == "all", yes = "", no = paste(" for", disease_)
       )
-    } else {
-      lapply(edist, validate_epidist)
-
-      message(
-        "Returning ", nrow(eparam), " results that match the criteria ",
-        "(", sum(is_param), " are parameterised). \n",
-        "Use subset to filter by entry variables or ",
-        "single_epidist to return a single entry. \n",
-        "To retrieve the short citation for each use the ",
-        "'get_citation' function"
+      epi_dist_str <- ifelse(test = epi_dist == "all", yes = "", no = epi_dist_)
+      stop(
+        epi_dist_str, " distribution not available", disease_str,
+        call. = FALSE
+      )
+    },
+    {
+      disease <- match.arg(
+        arg = disease,
+        choices = disease_db,
+        several.ok = FALSE
+      )
+      epi_dist <- match.arg(
+        arg = epi_dist,
+        choices = epi_dist_db,
+        several.ok = FALSE
       )
     }
+  )
+
+  # filter based on disease
+  if (disease != "all") {
+    multi_epidist <- Filter(
+      f = function(x) clean_disease(x$disease$disease) == disease,
+      x = multi_epidist
+    )
   }
 
-  # return epidist
-  edist
+  # filter by epi dist
+  if (epi_dist != "all") {
+    multi_epidist <- Filter(
+      f = function(x) clean_epi_dist(x$epi_dist) == epi_dist,
+      x = multi_epidist
+    )
+  }
+
+  # return multi_epidist
+  multi_epidist
+}
+
+#' Format data from JSON database into `<epidist>` objects
+#'
+#' @param x A single entry (element/object) from the database.
+#'
+#' @return An `<epidist>` object.
+#' @keywords internal
+#' @noRd
+.format_epidist <- function(x) {
+  # convert nulls to NA
+  null2na <- function(l) {
+    if (is.list(l)) l <- lapply(l, null2na) else if (is.null(l)) l <- NA
+    l
+  }
+  x <- lapply(x, null2na)
```

Comment:
Consider using `rapply()` instead which imo/iirc covers this use case.

Reply:
I take this suggestion back, as this is the exact use case where `rapply()` fails, [per this SO post.](https://stackoverflow.com/questions/38950005/how-to-manipulate-null-elements-in-a-nested-list/)